Components/Helper/CartRestore.php

```diff
@@ -0,0 +1,186 @@
+use Doctrine\ORM\OptimisticLockException;
+use Shopware\Models\Order\Basket;
+
+/**
+ *
+ * @package Shopware
+ * @subpackage Plugins_Frontend
+ */
+class Shopware_Plugins_Frontend_NostoTagging_Components_Helper_CartRestore
+{
+    const CART_RESTORE_URL_PARAMETER = 'h';
+
+    /**
+     * @param $hash
+     * @param $currentSessionId
+     */
+    public function restoreCart($hash, $currentSessionId)
+    {
+        $nostoCustomer = Shopware()
+            ->Models()
+            ->getRepository(CustomerModel::class)
+            ->findOneBy(array('restoreCartHash' => $hash));
+
+        $this->updateBasket($nostoCustomer->getSessionId(), $currentSessionId);
+    }
+
+    /**
+     * Wrapper function to return the current session ID
+     * @return null|string
+     */
+    public function getSessionId()
+    {
+        return Shopware()->Session()->get('sessionId');
+    }
+
+    /**
+     * Updates the sessions return a unique URL for the current cart
+     *
+     * @return string
+     * @throws OptimisticLockException
+     */
+    public function generateRestoreToCartLink()
+    {
+        $nostoCustomer = $this->updateNostoId();
+        if ($nostoCustomer && $nostoCustomer->getRestoreCartHash()) {
+            $basket = Shopware()
+                ->Models()
+                ->getRepository(Basket::class)
+                ->findOneBy(array('sessionId' => $nostoCustomer->getSessionId()));
+            if ($basket !== null) {
+                return NostoHelperUrl::generateRestoreCartUrl(
+                    Shopware()->Shop(),
+                    $nostoCustomer->getRestoreCartHash()
+                );
+            }
+        }
+        return '-';
+    }
+
+    /**
+     * Generate unique hash for restore cart
+     * Size of it equals to or less than restore_cart_hash column length
+     *
+     * @return string
+     */
+    public function generateRestoreCartHash()
+    {
+        $hash = hash(NostoCustomerComponent::VISITOR_HASH_ALGO, uniqid('nostocartrestore', true));
+        if (strlen($hash) > CustomerModel::NOSTO_TAGGING_RESTORE_CART_ATTRIBUTE_LENGTH) {
+            $hash = substr($hash, 0, CustomerModel::NOSTO_TAGGING_RESTORE_CART_ATTRIBUTE_LENGTH);
+        }
+        return $hash;
+    }
+
+    /**
+     * Update basket session
+     *
+     * @param $sessionId
+     * @param $newSessionId
+     * @return bool
+     */
+    public function updateBasket($sessionId, $newSessionId)
+    {
+        try {
+            $basketItems = Shopware()
+                ->Models()
+                ->getRepository(Basket::class)
+                ->findBy(array('sessionId' => $sessionId));
+
+            foreach ($basketItems as $basketItem) {
+                /** @var Basket $basketItem */
+                $basketItem->setSessionId($newSessionId);
+                Shopware()->Models()->persist($basketItem);
+                Shopware()->Models()->flush($basketItem);
+            }
```

Comment:
We need to update also the sessionId or corresponding cart hash for Nosto Customer (`s_nostotagging_cusromer`). Otherwise you can only restore the cart once using the link. It should work so that the link can be clicked multiple times and the cart would be restored. 

Please note that there's a unique index for sessionId in the table so it might be easier to just update restore cart hash with matching sessionId (the old session id).

Reply:
Fixed, thanks